File: mergesort.go
package main

import (
	"fmt"
	"math/rand"
)

func merge(a []int, b []int) []int {
	var r = make([]int, len(a)+len(b))
	var i = 0
	var j = 0

	for i < len(a) && j < len(b) {
		if a[i] <= b[j] {
			r[i+j] = a[i]
			i++
		} else {
			r[i+j] = b[j]
			j++
		}
	}

	for i < len(a) {
		r[i+j] = a[i]
		i++
	}
	for j < len(b) {
		r[i+j] = b[j]
		j++
	}

	return r
}

func Mergesort(items []int) chan []int {
	out := make(chan []int)
	go func() {
		if len(items) < 2 {
			out <- items
			return
		}

		var middle = len(items) / 2

		left := Mergesort(items[:middle])
		right := Mergesort(items[middle:])

		a := <-left
		b := <-right

		out <- merge(a, b)
	}()
	return out
}

func randIntArray(length int) []int {
	input := make([]int, length)
	for i := 0; i < length; i += 1 {
		input[i] = rand.Int()
	}
	return input
}

func main() {
	input := randIntArray(10000)
	result := Mergesort(input)
	solution := <-result
	fmt.Println(solution)
}

File: mergesort.hs
import Control.Concurrent
import System.Random

fsthalf :: [a] -> [a]
fsthalf xs = take (length xs `div` 2) xs

sndhalf :: [a] -> [a]
sndhalf xs = drop (length xs `div` 2) xs

merge :: Ord a => [a] -> [a] -> [a]
merge xs [] = xs
merge [] ys = ys
merge (x:xs) (y:ys) 
          | (x <= y)  = x:(merge xs (y:ys)) 
          | otherwise = y:(merge (x:xs) ys)


mergesort :: Ord a => [a] -> MVar [a] -> IO()
mergesort [] ch = putMVar ch [] 
mergesort [x] ch = putMVar ch [x]

mergesort xs ch = do
    left <- newEmptyMVar
    right <- newEmptyMVar
    forkIO $ mergesort (fsthalf xs) left
    forkIO $ mergesort (sndhalf xs) right
    a <- takeMVar left
    b <- takeMVar right
    putMVar ch (merge a b)

main = do
    g <- getStdGen
    let list =  take 10000 (randoms g :: [Int])
    result <- newEmptyMVar
    forkIO $ mergesort list result
    sorted <- takeMVar result
    print(sorted)